File: tests/python/test_updated_bindings.py
#!/usr/bin/env python3
"""
Test script for updated QtForge Python bindings
Tests all new functionality added to the bindings.
"""

import sys
import traceback

def test_core_bindings() -> None:
    """Test core module bindings including new interfaces."""
    print("Testing core bindings...")

    try:
        import qtforge.core as core

        # Test enums
        print("  Testing enums...")
        assert hasattr(core, 'PluginState')
        assert hasattr(core, 'PluginCapability')
        assert hasattr(core, 'PluginPriority')
        assert hasattr(core, 'PluginLifecycleEvent')
        assert hasattr(core, 'PluginType')

        # Test new PluginType enum
        assert hasattr(core.PluginType, 'Native')
        assert hasattr(core.PluginType, 'Python')
        assert hasattr(core.PluginType, 'Lua')
        assert hasattr(core.PluginType, 'Remote')
        assert hasattr(core.PluginType, 'Composite')

        # Test core classes
        print("  Testing core classes...")
        assert hasattr(core, 'IPlugin')
        assert hasattr(core, 'PluginManager')
        assert hasattr(core, 'PluginLoader')
        assert hasattr(core, 'PluginRegistry')
        assert hasattr(core, 'PluginDependencyResolver')
        assert hasattr(core, 'PluginLifecycleManager')

        # Test new interface classes
        print("  Testing new interface classes...")
        assert hasattr(core, 'InterfaceCapability')
        assert hasattr(core, 'InterfaceDescriptor')
        assert hasattr(core, 'IAdvancedPlugin')
        assert hasattr(core, 'IDynamicPlugin')
        assert hasattr(core, 'IServicePlugin')

        # Test service-related enums
        print("  Testing service enums...")
        assert hasattr(core, 'ServiceExecutionMode')
        assert hasattr(core, 'ServiceState')
        assert hasattr(core, 'ServicePriority')
        assert hasattr(core, 'ServiceHealth')

        # Test service enum values
        assert hasattr(core.ServiceExecutionMode, 'MainThread')
        assert hasattr(core.ServiceExecutionMode, 'WorkerThread')
        assert hasattr(core.ServiceState, 'Running')
        assert hasattr(core.ServiceState, 'Stopped')
        assert hasattr(core.ServicePriority, 'Normal')
        assert hasattr(core.ServicePriority, 'High')
        assert hasattr(core.ServiceHealth, 'Healthy')
        assert hasattr(core.ServiceHealth, 'Unhealthy')

        # Test factory functions
        print("  Testing factory functions...")
        manager = core.create_plugin_manager()
        assert manager is not None

        loader = core.create_plugin_loader()
        assert loader is not None

        registry = core.create_plugin_registry()
        assert registry is not None

        resolver = core.create_plugin_dependency_resolver()
        assert resolver is not None

        lifecycle_manager = core.create_plugin_lifecycle_manager()
        assert lifecycle_manager is not None

        print("  Core bindings test passed!")
        return True

    except Exception as e:
        print(f"  Core bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_communication_bindings() -> None:
    """Test communication module bindings including service contracts."""
    print("Testing communication bindings...")

    try:
        import qtforge.communication as comm

        # Test enums
        print("  Testing communication enums...")
        assert hasattr(comm, 'DeliveryMode')
        assert hasattr(comm, 'MessagePriority')
        assert hasattr(comm, 'ServiceCapability')

        # Test service capability enum values
        assert hasattr(comm.ServiceCapability, 'Synchronous')
        assert hasattr(comm.ServiceCapability, 'Asynchronous')
        assert hasattr(comm.ServiceCapability, 'Streaming')
        assert hasattr(comm.ServiceCapability, 'Transactional')

        # Test message classes
        print("  Testing message classes...")
        assert hasattr(comm, 'IMessage')
        assert hasattr(comm, 'BasicMessage')
        assert hasattr(comm, 'MessageBus')

        # Test service contract classes
        print("  Testing service contract classes...")
        assert hasattr(comm, 'ServiceVersion')
        assert hasattr(comm, 'ServiceMethodDescriptor')
        assert hasattr(comm, 'ServiceContract')

        # Test factory functions
        print("  Testing communication factory functions...")
        message_bus = comm.create_message_bus()
        assert message_bus is not None

        print("  Communication bindings test passed!")
        return True

    except Exception as e:
        print(f"  Communication bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_security_bindings() -> None:
    """Test security module bindings including new enums."""
    print("Testing security bindings...")

    try:
        import qtforge.security as security

        # Test enums
        print("  Testing security enums...")
        assert hasattr(security, 'SecurityLevel')
        assert hasattr(security, 'PluginPermission')
        assert hasattr(security, 'TrustLevel')

        # Test permission enum values
        assert hasattr(security.PluginPermission, 'FileSystemRead')
        assert hasattr(security.PluginPermission, 'FileSystemWrite')
        assert hasattr(security.PluginPermission, 'NetworkAccess')
        assert hasattr(security.PluginPermission, 'ProcessCreation')

        # Test trust level enum values
        assert hasattr(security.TrustLevel, 'Untrusted')
        assert hasattr(security.TrustLevel, 'Limited')
        assert hasattr(security.TrustLevel, 'Trusted')
        assert hasattr(security.TrustLevel, 'FullyTrusted')

        # Test security classes
        print("  Testing security classes...")
        assert hasattr(security, 'SecurityManager')
        assert hasattr(security, 'SecurityValidator')
        assert hasattr(security, 'SignatureVerifier')
        assert hasattr(security, 'PermissionManager')
        assert hasattr(security, 'SecurityPolicyEngine')

        # Test factory functions
        print("  Testing security factory functions...")
        manager = security.create_security_manager()
        assert manager is not None

        validator = security.create_security_validator()
        assert validator is not None

        verifier = security.create_signature_verifier()
        assert verifier is not None

        perm_manager = security.create_permission_manager()
        assert perm_manager is not None

        policy_engine = security.create_security_policy_engine()
        assert policy_engine is not None

        print("  Security bindings test passed!")
        return True

    except Exception as e:
        print(f"  Security bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_managers_bindings() -> None:
    """Test managers module bindings."""
    print("Testing managers bindings...")

    try:
        import qtforge.managers as managers

        # Test enums
        print("  Testing manager enums...")
        assert hasattr(managers, 'ConfigurationScope')
        assert hasattr(managers, 'ConfigurationChangeType')
        assert hasattr(managers, 'LogLevel')

        # Test manager classes
        print("  Testing manager classes...")
        assert hasattr(managers, 'ConfigurationManager')
        assert hasattr(managers, 'LoggingManager')
        assert hasattr(managers, 'ResourceManager')
        assert hasattr(managers, 'PluginVersionManager')

        # Test factory functions
        print("  Testing manager factory functions...")
        config_manager = managers.create_configuration_manager()
        assert config_manager is not None

        logging_manager = managers.create_logging_manager()
        assert logging_manager is not None

        resource_manager = managers.create_resource_manager()
        assert resource_manager is not None

        print("  Managers bindings test passed!")
        return True

    except Exception as e:
        print(f"  Managers bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_monitoring_bindings() -> None:
    """Test monitoring module bindings."""
    print("Testing monitoring bindings...")

    try:
        import qtforge.monitoring as monitoring

        # Test monitoring classes
        print("  Testing monitoring classes...")
        assert hasattr(monitoring, 'PluginHotReloadManager')
        assert hasattr(monitoring, 'PluginMetricsCollector')

        # Test factory functions
        print("  Testing monitoring factory functions...")
        hot_reload_manager = monitoring.create_hot_reload_manager()
        assert hot_reload_manager is not None

        metrics_collector = monitoring.create_metrics_collector()
        assert metrics_collector is not None

        print("  Monitoring bindings test passed!")
        return True

    except Exception as e:
        print(f"  Monitoring bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_transactions_bindings() -> None:
    """Test transactions module bindings."""
    print("Testing transactions bindings...")

    try:
        import qtforge.transactions as transactions

        # Test transaction enums
        print("  Testing transaction enums...")
        assert hasattr(transactions, 'TransactionState')
        assert hasattr(transactions, 'IsolationLevel')

        # Test transaction classes
        print("  Testing transaction classes...")
        assert hasattr(transactions, 'PluginTransactionManager')

        # Test singleton access
        print("  Testing transaction manager singleton...")
        manager = transactions.get_transaction_manager()
        assert manager is not None

        print("  Transactions bindings test passed!")
        return True

    except Exception as e:
        print(f"  Transactions bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_composition_bindings() -> None:
    """Test composition module bindings."""
    print("Testing composition bindings...")

    try:
        import qtforge.composition as composition

        # Test composition enums
        print("  Testing composition enums...")
        assert hasattr(composition, 'CompositionStrategy')
        assert hasattr(composition, 'PluginRole')

        # Test composition classes
        print("  Testing composition classes...")
        assert hasattr(composition, 'CompositionManager')

        # Test singleton access
        print("  Testing composition manager singleton...")
        manager = composition.get_composition_manager()
        assert manager is not None

        print("  Composition bindings test passed!")
        return True

    except Exception as e:
        print(f"  Composition bindings test failed: {e}")
        traceback.print_exc()
        return False

def test_marketplace_bindings() -> None:
    """Test marketplace module bindings."""
    print("Testing marketplace bindings...")

    try:
        import qtforge.marketplace as marketplace

        # Test marketplace classes
        print("  Testing marketplace classes...")
        assert hasattr(marketplace, 'PluginMarketplace')

        print("  Marketplace bindings test passed!")
        return True

    except Exception as e:
        print(f"  Marketplace bindings test failed: {e}")
        traceback.print_exc()
        return False

def main() -> None:
    """Run all binding tests."""
    print("QtForge Python Bindings Test Suite")
    print("=" * 40)

    tests = [
        test_core_bindings,
        test_communication_bindings,
        test_security_bindings,
        test_managers_bindings,
        test_monitoring_bindings,
        test_transactions_bindings,
        test_composition_bindings,
        test_marketplace_bindings,
    ]

    passed = 0
    failed = 0

    for test in tests:
        try:
            if test():
                passed += 1
            else:
                failed += 1
        except Exception as e:
            print(f"Test {test.__name__} crashed: {e}")
            failed += 1
        print()

    print("=" * 40)
    print(f"Test Results: {passed} passed, {failed} failed")

    if failed == 0:
        print("All tests passed!")
        return 0
    else:
        print("Some tests failed!")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: tests/lua/test_updated_bindings.lua
#!/usr/bin/env lua
--[[
Test script for updated QtForge Lua bindings
Tests all new functionality added to the bindings.
--]]

-- Test helper functions
local function assert_not_nil(value, message)
    if value == nil then
        error(message or "Value should not be nil")
    end
end

local function assert_type(value, expected_type, message)
    if type(value) ~= expected_type then
        error(message or string.format("Expected %s, got %s", expected_type, type(value)))
    end
end

local function test_core_bindings()
    print("Testing core bindings...")

    local success, err = pcall(function()
        -- Test that qtforge module is available
        assert_not_nil(qtforge, "qtforge module should be available")
        assert_not_nil(qtforge.core, "qtforge.core module should be available")

        -- Test enums
        print("  Testing enums...")
        assert_not_nil(PluginState, "PluginState enum should be available")
        assert_not_nil(PluginCapability, "PluginCapability enum should be available")
        assert_not_nil(PluginPriority, "PluginPriority enum should be available")
        assert_not_nil(PluginType, "PluginType enum should be available")
        assert_not_nil(ServiceExecutionMode, "ServiceExecutionMode enum should be available")
        assert_not_nil(ServiceState, "ServiceState enum should be available")
        assert_not_nil(ServicePriority, "ServicePriority enum should be available")
        assert_not_nil(ServiceHealth, "ServiceHealth enum should be available")

        -- Test enum values
        assert_not_nil(PluginState.Unloaded, "PluginState.Unloaded should be available")
        assert_not_nil(PluginState.Loaded, "PluginState.Loaded should be available")
        assert_not_nil(PluginState.Running, "PluginState.Running should be available")

        assert_not_nil(PluginCapability.Service, "PluginCapability.Service should be available")
        assert_not_nil(PluginCapability.Network, "PluginCapability.Network should be available")

        assert_not_nil(PluginPriority.Normal, "PluginPriority.Normal should be available")
        assert_not_nil(PluginPriority.High, "PluginPriority.High should be available")

        -- Test factory functions
        print("  Testing factory functions...")
        local manager = qtforge.core.create_plugin_manager()
        assert_not_nil(manager, "create_plugin_manager should return a manager")

        local loader = qtforge.core.create_plugin_loader()
        assert_not_nil(loader, "create_plugin_loader should return a loader")

        local registry = qtforge.core.create_plugin_registry()
        assert_not_nil(registry, "create_plugin_registry should return a registry")

        local resolver = qtforge.core.create_plugin_dependency_resolver()
        assert_not_nil(resolver, "create_plugin_dependency_resolver should return a resolver")

        -- Test utility functions
        print("  Testing utility functions...")
        local version = qtforge.core.version(1, 2, 3)
        assert_not_nil(version, "version constructor should work")

        local error_obj = qtforge.core.error(PluginErrorCode.LoadFailed, "Test error")
        assert_not_nil(error_obj, "error constructor should work")

        local metadata = qtforge.core.metadata()
        assert_not_nil(metadata, "metadata constructor should work")
    end)

    if success then
        print("  Core bindings test passed!")
        return true
    else
        print("  Core bindings test failed: " .. tostring(err))
        return false
    end
end

local function test_managers_bindings()
    print("Testing managers bindings...")

    local success, err = pcall(function()
        -- Test that managers module is available
        assert_not_nil(qtforge.managers, "qtforge.managers module should be available")

        -- Test configuration scope enum
        print("  Testing configuration enums...")
        assert_not_nil(ConfigurationScope, "ConfigurationScope enum should be available")
        assert_not_nil(ConfigurationScope.Global, "ConfigurationScope.Global should be available")
        assert_not_nil(ConfigurationScope.Plugin, "ConfigurationScope.Plugin should be available")
        assert_not_nil(ConfigurationScope.User, "ConfigurationScope.User should be available")

        -- Test logging level enum
        assert_not_nil(LogLevel, "LogLevel enum should be available")
        assert_not_nil(LogLevel.Debug, "LogLevel.Debug should be available")
        assert_not_nil(LogLevel.Info, "LogLevel.Info should be available")
        assert_not_nil(LogLevel.Warning, "LogLevel.Warning should be available")
        assert_not_nil(LogLevel.Error, "LogLevel.Error should be available")

        -- Test factory functions (if available)
        print("  Testing manager factory functions...")
        if qtforge.managers.create_configuration_manager then
            local config_manager = qtforge.managers.create_configuration_manager()
            assert_not_nil(config_manager, "create_configuration_manager should return a manager")
        end

        if qtforge.managers.create_logging_manager then
            local logging_manager = qtforge.managers.create_logging_manager()
            assert_not_nil(logging_manager, "create_logging_manager should return a manager")
        end

        if qtforge.managers.create_resource_manager then
            local resource_manager = qtforge.managers.create_resource_manager()
            assert_not_nil(resource_manager, "create_resource_manager should return a manager")
        end
    end)

    if success then
        print("  Managers bindings test passed!")
        return true
    else
        print("  Managers bindings test failed: " .. tostring(err))
        return false
    end
end

local function test_security_bindings()
    print("Testing security bindings...")

    local success, err = pcall(function()
        -- Test that security module is available
        assert_not_nil(qtforge.security, "qtforge.security module should be available")

        -- Test security level enum
        print("  Testing security enums...")
        assert_not_nil(SecurityLevel, "SecurityLevel enum should be available")
        assert_not_nil(SecurityLevel.None, "SecurityLevel.None should be available")
        assert_not_nil(SecurityLevel.Low, "SecurityLevel.Low should be available")
        assert_not_nil(SecurityLevel.Medium, "SecurityLevel.Medium should be available")
        assert_not_nil(SecurityLevel.High, "SecurityLevel.High should be available")
        assert_not_nil(SecurityLevel.Maximum, "SecurityLevel.Maximum should be available")

        -- Test trust level enum
        assert_not_nil(TrustLevel, "TrustLevel enum should be available")
        assert_not_nil(TrustLevel.Untrusted, "TrustLevel.Untrusted should be available")
        assert_not_nil(TrustLevel.Limited, "TrustLevel.Limited should be available")
        assert_not_nil(TrustLevel.Trusted, "TrustLevel.Trusted should be available")
        assert_not_nil(TrustLevel.FullyTrusted, "TrustLevel.FullyTrusted should be available")

        -- Test factory functions (if available)
        print("  Testing security factory functions...")
        if qtforge.security.create_security_manager then
            local security_manager = qtforge.security.create_security_manager()
            assert_not_nil(security_manager, "create_security_manager should return a manager")
        end

        if qtforge.security.create_security_validator then
            local validator = qtforge.security.create_security_validator()
            assert_not_nil(validator, "create_security_validator should return a validator")
        end
    end)

    if success then
        print("  Security bindings test passed!")
        return true
    else
        print("  Security bindings test failed: " .. tostring(err))
        return false
    end
end

local function test_orchestration_bindings()
    print("Testing orchestration bindings...")

    local success, err = pcall(function()
        -- Test that orchestration module is available
        assert_not_nil(qtforge.orchestration, "qtforge.orchestration module should be available")

        -- Test step status enum
        print("  Testing orchestration enums...")
        if StepStatus then
            assert_not_nil(StepStatus.Pending, "StepStatus.Pending should be available")
            assert_not_nil(StepStatus.Running, "StepStatus.Running should be available")
            assert_not_nil(StepStatus.Completed, "StepStatus.Completed should be available")
            assert_not_nil(StepStatus.Failed, "StepStatus.Failed should be available")
        end

        -- Test execution mode enum
        if ExecutionMode then
            assert_not_nil(ExecutionMode.Sequential, "ExecutionMode.Sequential should be available")
            assert_not_nil(ExecutionMode.Parallel, "ExecutionMode.Parallel should be available")
            assert_not_nil(ExecutionMode.Pipeline, "ExecutionMode.Pipeline should be available")
        end

        -- Test factory functions
        print("  Testing orchestration factory functions...")
        if qtforge.orchestration.create_workflow_step then
            local step = qtforge.orchestration.create_workflow_step("test_id", "Test Step", "test_plugin")
            assert_not_nil(step, "create_workflow_step should return a step")
        end
    end)

    if success then
        print("  Orchestration bindings test passed!")
        return true
    else
        print("  Orchestration bindings test failed: " .. tostring(err))
        return false
    end
end

local function test_monitoring_bindings()
    print("Testing monitoring bindings...")

    local success, err = pcall(function()
        -- Test that monitoring module is available
        assert_not_nil(qtforge.monitoring, "qtforge.monitoring module should be available")

        -- Test factory functions (if available)
        print("  Testing monitoring factory functions...")
        if qtforge.monitoring.create_hot_reload_manager then
            local hot_reload_manager = qtforge.monitoring.create_hot_reload_manager()
            assert_not_nil(hot_reload_manager, "create_hot_reload_manager should return a manager")
        end

        if qtforge.monitoring.create_metrics_collector then
            local metrics_collector = qtforge.monitoring.create_metrics_collector()
            assert_not_nil(metrics_collector, "create_metrics_collector should return a collector")
        end
    end)

    if success then
        print("  Monitoring bindings test passed!")
        return true
    else
        print("  Monitoring bindings test failed: " .. tostring(err))
        return false
    end
end

local function test_transactions_bindings()
    print("Testing transactions bindings...")

    local success, err = pcall(function()
        -- Test that transactions module is available
        assert_not_nil(qtforge.transactions, "qtforge.transactions module should be available")

        -- Test transaction enums
        print("  Testing transaction enums...")
        assert_not_nil(TransactionState, "TransactionState enum should be available")
        assert_not_nil(IsolationLevel, "IsolationLevel enum should be available")

        -- Test transaction manager
        print("  Testing transaction manager...")
        local manager = qtforge.transactions.get_transaction_manager()
        assert_not_nil(manager, "get_transaction_manager should return a manager")

        print("  Transactions bindings test passed!")
        return true
    end)

    if not success then
        print("  Transactions bindings test failed: " .. tostring(err))
        return false
    end

    return success
end

local function test_composition_bindings()
    print("Testing composition bindings...")

    local success, err = pcall(function()
        -- Test that composition module is available
        assert_not_nil(qtforge.composition, "qtforge.composition module should be available")

        -- Test composition enums
        print("  Testing composition enums...")
        assert_not_nil(CompositionStrategy, "CompositionStrategy enum should be available")
        assert_not_nil(PluginRole, "PluginRole enum should be available")

        -- Test composition manager
        print("  Testing composition manager...")
        local manager = qtforge.composition.get_composition_manager()
        assert_not_nil(manager, "get_composition_manager should return a manager")

        print("  Composition bindings test passed!")
        return true
    end)

    if not success then
        print("  Composition bindings test failed: " .. tostring(err))
        return false
    end

    return success
end

local function test_marketplace_bindings()
    print("Testing marketplace bindings...")

    local success, err = pcall(function()
        -- Test that marketplace module is available
        assert_not_nil(qtforge.marketplace, "qtforge.marketplace module should be available")

        -- Test marketplace classes
        print("  Testing marketplace classes...")
        assert_not_nil(PluginMarketplace, "PluginMarketplace class should be available")

        print("  Marketplace bindings test passed!")
        return true
    end)

    if not success then
        print("  Marketplace bindings test failed: " .. tostring(err))
        return false
    end

    return success
end

-- Main test function
local function main()
    print("QtForge Lua Bindings Test Suite")
    print(string.rep("=", 40))

    local tests = {
        test_core_bindings,
        test_managers_bindings,
        test_security_bindings,
        test_orchestration_bindings,
        test_monitoring_bindings,
        test_transactions_bindings,
        test_composition_bindings,
        test_marketplace_bindings,
    }

    local passed = 0
    local failed = 0

    for _, test in ipairs(tests) do
        local success = test()
        if success then
            passed = passed + 1
        else
            failed = failed + 1
        end
        print()
    end

    print(string.rep("=", 40))
    print(string.format("Test Results: %d passed, %d failed", passed, failed))

    if failed == 0 then
        print("All tests passed!")
        return 0
    else
        print("Some tests failed!")
        return 1
    end
end

-- Run the tests
return main()
